enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,449 @@
+If any of those criterias are met for longer than given timeouts, the controller
+creates a remediation CR according to the template in the NHC CRD.
+Timeouts will be defined as part of the NHC CRD .
+
+### Implementation Details
+
+#### NodeHealthCheck CRD
+- Enable watching a node based on a label selector. 
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further action will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+E.g:
+- Create a remedy CR when node is `ready=false` or `ready=Unknown` condition for more than 5m.
+- I want to avoid triggering remediation if 40% or more of the targets of this pool are unhealthy at the same time.
+
+
+```yaml
+apiVersion: remediation.medik8s.io/v1alpha1
+kind: NodeHealthCheck
+metadata:
+  name: example
+  # this CRD is cluster-scope, namespace is not needed  
+spec:
+  maxUnhealthy: "49%"
+  selector:
+    matchExpressions:
+      key: node-role.kubernetes.io/worker
+      operator: exists
+  unhealthyConditions:
+    - type: "Ready"
+      status:  "Unknown"
+      duration: "300s"
+   - type: "Ready"
+     status:  "False"
+     duration: "300s"
+  remediationTemplate:
+    kind: PoisonPillRemediationTemplate
+    apiVersion: poison-pill.medik8s.io/v1alphaX
+    name: M3_REMEDIATION_GROUP
+status:
+  healthyNodes: 5
+  obeservedNodes: 5
+```
+
+#### NodeHealthCheck controller
+Watch:
+- Watch NodeHealthCheck resources
+- Watch nodes with an event handler e.g controller runtime `EnqueueRequestsFromMapFunc` which returns NodeHealthCheck resources.
+
+Reconcile:
+- Periodically enqueue a reconcile request
+- Fetch all nodes targets. E.g:
+
+
+- Calculate the number of unhealthy targets.
+- Compare current number against `maxUnhealthy` threshold and temporary short
+ circuits if the threshold is met.
+- Create remedy objects for unhealthy targets according to the template in NHC object
+
+### Notes/Constraints
+
+The functionality of this controller have a lot in common with the Machine Health
+Check controller except that its creating objects and not
+performing any actions per se. When deploying this controller make
+sure 1. you have a controller implementing the API, e.g [poison pill]
+2. the machine health check controller is not running, or in case of a mixed
+ the node health checks selector avoid targeting nodes with machines.
+
+
+### Risks and Mitigations
+
+#### Remediation Interaction with ClusterAutoScalar
+
+Misconfiguration of nodes provisioning can cause nodes to fail to get operational.
+Remediation is blind to such problems and may interfere with diagnosing the problem.
+The backoff mechanism mentioned in MHC may be suited for Machines, because they
+have the MachineSet name that can be recorded and tracked. Without the machine API
```

Comment:
I don't think MHC knows anything about MachineSets 🤔  I also don't think we have a backoff mechanism, what is this referring to?